src/main/java/co/rsk/federate/signing/hsm/advanceblockchain/HsmBookkeepingClientImpl.java

```diff
@@ -138,26 +100,100 @@ private void addBlocksToPayload(ObjectNode payload, String[] blockHeaderChunk) {
+    private void addBrothersToPayload(ObjectNode payload, List<String[]> brothers) {
         final String BROTHERS_FIELD = "brothers";
         ArrayNode brothersFieldData = new ObjectMapper().createArrayNode();
-        for (String blockHeader : blockHeaderChunk) {
-            // TODO: This is currently sending empty arrays as brothers to the HSM V3 for compatibility with V2
-            //  This should be changed to sending the actual brothers when HSM V3 is fully implemented
-            brothersFieldData.add(new ObjectMapper().createArrayNode());
+        for (String[] blockHeaderBrothers : brothers) {
+            ArrayNode fieldData = new ObjectMapper().createArrayNode();
+            for (String brother : blockHeaderBrothers) {
+                fieldData.add(brother);
+            }
+            brothersFieldData.add(fieldData);
         }
         payload.set(BROTHERS_FIELD, brothersFieldData);
     }
 
+    private List<String[]> getBrothers(String[] blockHeaderChunk, AdvanceBlockchainMessage message)
+        throws HSMBlockchainBookkeepingRelatedException {
+        List<String[]> brothers = new ArrayList<>();
+        for (String blockHeader : blockHeaderChunk) {
+            brothers.add(message.getParsedBrothers(blockHeader));
+        }
+        return brothers;
+    }
+
     @Override
     public void updateAncestorBlock(UpdateAncestorBlockMessage updateAncestorBlockMessage) throws HSMClientException {
-        sendBlockHeadersChunks(updateAncestorBlockMessage.getData(), "updateAncestorBlock", true);
+        final String UPDATE_ANCESTOR_METHOD_NAME = "updateAncestorBlock";
+        List<String> blockHeaders = updateAncestorBlockMessage.getData();
+        List<String[]> blockHeadersChunks = new ArrayList<>();
+        int chunkIndex = 0;
+        try {
+            validateBlockHeaders(blockHeaders, UPDATE_ANCESTOR_METHOD_NAME);
+            blockHeadersChunks = getChunks(blockHeaders.toArray(new String[]{}), maxChunkSize, true);
+
+            logger.trace("[updateAncestorBlock] Going to send {} headers in {} chunks.", blockHeaders.size(), blockHeadersChunks.size());
+            for (int i = 0; i < blockHeadersChunks.size(); i++) {
+                String[] blockHeaderChunk = blockHeadersChunks.get(i);
+                ObjectNode payload = this.hsmClientProtocol.buildCommand(UPDATE_ANCESTOR_METHOD_NAME, getVersion());
+                addBlocksToPayload(payload, blockHeaderChunk);
+
+                if (isStopped) {
+                    return;
+                }
+                chunkIndex = i + 1;
+                logger.trace("[updateAncestorBlock] chunk {}/{}", chunkIndex, blockHeadersChunks.size());
+                this.hsmClientProtocol.send(payload);
+            }
+        } catch (HSMClientException e) {
+            logger.warn(
+                "[updateAncestorBlock] failed sending {}/{} chunks. Error: {}",
+                chunkIndex,
+                blockHeadersChunks.size(),
```

Comment:
When `validateBlockHeaders` throws an error, It will always be `0`. What if you call `getChunks` out of the try/catch as it was before? Considering the message of the log, neither an exception from `validateBlockHeaders` nor `getChunks` is related to the action of sending chunks(from my understanding, correct if I'm wrong).

So, I think neither `validateBlockHeaders` nor `blockHeadersChunks` should be part of the try/catch. Those operations should be done out of it.

Reply:
The Log Is For Information Sake, The try/catch Aim Is To Throw The Exception To The Caller.
**validateBlockHeaders** Throws Some Couple Of Exceptions That Needs To Be Caught And Thrown To The Caller (line 155) , Hence It Can't Be Outside.